### ISSUE TITLE
systemd-timesyncd is not installed on oel76 instance, yet Scylla did not install ntp or chrony services as a replacement

### ISSUE THREAD

**Issue by yarongilor** [scylladb/scylla#10608]
## Installation details

Kernel Version: 4.14.35-1902.5.2.2.el7uek.x86_64

Cluster size: 1 nodes (i3.large)

Scylla Nodes used in this run:
- artifacts-oel76-jenkins-db-node-e3f8efcf-1 (34.244.161.148 | 10.0.2.48) (shards: 2)

OS / Image: `ami-0d6a24fe35fdf5dc4` (aws: eu-west-1)

Test: `artifacts-oel76-test`
Test id: `e3f8efcf-d430-409b-9ff5-af2504d39386`
Test name: `scylla-5.0/artifacts-offline-install/artifacts-oel76-test`
Test config file(s):

- [oel76.yaml](https://github.com/scylladb/scylla-cluster-tests/blob/598a6dc92f57e9f386a4280d92e73128151cafb5/test-cases/artifacts/oel76.yaml)

## Issue description

**>>>>>>>**
systemd-timesyncd is not installed on oel76 instance during artifacts test, yet Scylla did not install ntp or chrony services as a replacement

SCT error log:
```
< t:2022-05-17 21:09:18,057 f:remote_base.py  l:520  c:RemoteLibSSH2CmdRunner p:DEBUG > Running command "systemctl list-units --full | grep -Fq "systemd-timesyncd""...
< t:2022-05-17 21:09:18,598 f:remote_base.py  l:520  c:RemoteLibSSH2CmdRunner p:DEBUG > Running command "systemctl list-units --full | grep -Fq "ntpd""...
< t:2022-05-17 21:09:19,141 f:remote_base.py  l:520  c:RemoteLibSSH2CmdRunner p:DEBUG > Running command "systemctl list-units --full | grep -Fq "chronyd""...
< t:2022-05-17 21:17:00,633 f:file_logger.py  l:101  c:sdcm.sct_events.file_logger p:INFO  > AssertionError: systemd-timesyncd is not installed on the instance, yet Scylla did not install ntp or chrony services as a replacement
```
```

2022-05-17 15:25:33.738: (TestFrameworkEvent Severity.ERROR) period_type=one-time event_id=3ae9178a-9f7b-4ffe-a71c-7c366ffa1947, source=ArtifactsTest.test_scylla_service (artifacts_test.ArtifactsTest) [check if scylla unnecessarily installed a time synchronization service]() message=Traceback (most recent call last):
File "/home/jenkins/slave/workspace/scylla-5.0/artifacts-offline-install/artifacts-oel76-test/scylla-cluster-tests/artifacts_test.py", line 325, in test_scylla_service
assert is_ntp_service_installed or is_chrony_service_installed, \
AssertionError: systemd-timesyncd is not installed on the instance, yet Scylla did not install ntp or chrony services as a replacement
```
**<<<<<<<**

- Restore Monitor Stack command: `$ hydra investigate show-monitor e3f8efcf-d430-409b-9ff5-af2504d39386`
- Restore monitor on AWS instance using [Jenkins job](https://jenkins.scylladb.com/view/QA/job/QA-tools/job/hydra-show-monitor/parambuild/?test_id=e3f8efcf-d430-409b-9ff5-af2504d39386)
- Show all stored logs command: `$ hydra investigate show-logs e3f8efcf-d430-409b-9ff5-af2504d39386`

## Logs:
- **db-cluster-e3f8efcf.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/e3f8efcf-d430-409b-9ff5-af2504d39386/20220517_152624/db-cluster-e3f8efcf.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/e3f8efcf-d430-409b-9ff5-af2504d39386/20220517_152624/db-cluster-e3f8efcf.tar.gz)
- **sct-runner-e3f8efcf.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/e3f8efcf-d430-409b-9ff5-af2504d39386/20220517_152624/sct-runner-e3f8efcf.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/e3f8efcf-d430-409b-9ff5-af2504d39386/20220517_152624/sct-runner-e3f8efcf.tar.gz)

[Jenkins job URL](https://jenkins.scylladb.com/job/scylla-5.0/job/artifacts-offline-install/job/artifacts-oel76-test/3/)

**Comment by roydahan** [comment]
@yarongilor how does OEL76 passes in regular mode (non-offline)?
Maybe it's something we don't provide as offline package and we should assume it's there...

**Comment by roydahan** [comment]
@yarongilor ping

**Comment by fgelcer** [comment]
ping @yarongilor 

please confirm what differs between the offline install one, and the regular artifacts

**Comment by yarongilor** [comment]
In non offline the package does exist, in offline mode it doesn't exist:
```
ubuntu@ip-10-0-1-226:~/sct-results$ ssh -i ~/.ssh/scylla-qa-ec2 clckwrk@63.35.234.251 "systemctl list-unit-files | egrep -i 'chrony'"
chrony-dnssrv@.service                        static  
chrony-wait.service                           disabled
chronyd.service                               enabled 
chrony-dnssrv@.timer                          disabled
ubuntu@ip-10-0-1-226:~/sct-results$ ssh -i ~/.ssh/scylla-qa-ec2 clckwrk@34.246.178.163 "systemctl list-unit-files | egrep -i 'chrony'"
ubuntu@ip-10-0-1-226:~/sct-results$ 
```

**Comment by avikivity** [comment]
We have to document that the user should install chronyd before starting.

**Comment by fgelcer** [comment]
> We have to document that the user should install chronyd before starting.

seems like either [this document](https://amoskong.github.io/scylladb/2021/01/27/scylla-offline-install/) is not right, or something changed since... but it states:
```
All dependencies will be contained in the installer
```

**Comment by fgelcer** [comment]
> In non offline the package does exist, in offline mode it doesn't exist:
> 
> ```
> ubuntu@ip-10-0-1-226:~/sct-results$ ssh -i ~/.ssh/scylla-qa-ec2 clckwrk@63.35.234.251 "systemctl list-unit-files | egrep -i 'chrony'"
> chrony-dnssrv@.service                        static  
> chrony-wait.service                           disabled
> chronyd.service                               enabled 
> chrony-dnssrv@.timer                          disabled
> ubuntu@ip-10-0-1-226:~/sct-results$ ssh -i ~/.ssh/scylla-qa-ec2 clckwrk@34.246.178.163 "systemctl list-unit-files | egrep -i 'chrony'"
> ubuntu@ip-10-0-1-226:~/sct-results$ 
> ```

@yarongilor , please find where is the gap in SCT...
i see the following code:
```
            with self.subTest("check if scylla unnecessarily installed a time synchronization service"):
                # Checks https://github.com/scylladb/scylla/issues/8339
                # If the instance already has systemd-timesyncd
                is_timesyncd_service_installed = self.check_service_existence(service_name="systemd-timesyncd")
                # Do note: On Redhat based distributions the services are named ntpd and chronyd, while on debian based
                # distributions they're named ntp and chrony.
                if self.node.is_rhel_like():
                    is_ntp_service_installed = self.check_service_existence(service_name="ntpd")
                    is_chrony_service_installed = self.check_service_existence(service_name="chronyd")
                else:
                    is_ntp_service_installed = self.check_service_existence(service_name="ntp")
                    is_chrony_service_installed = self.check_service_existence(service_name="chrony")
                if is_timesyncd_service_installed:
                    assert not is_ntp_service_installed, \
                        "systemd-timesyncd is already installed, yet scylla installed ntp service on top of it"
                    assert not is_chrony_service_installed, \
                        "systemd-timesyncd is already installed, yet scylla installed chrony service on top of it"
                else:
                    assert is_ntp_service_installed or is_chrony_service_installed, \
                        "systemd-timesyncd is not installed on the instance, yet Scylla did not install ntp or " \
                        "chrony services as a replacement"
```

i understand that, if `systemd-timesyncd` is installed, then we expect Scylla to not install `chrony` nor `ntp`, but if `systemd-timesyncd` is not installed, then Scylla must provide either `chrony` or `ntp`...
check if we just failed in the 1st verification, and didn't move forward when expected that, we might not have it installed.

**Comment by yarongilor** [comment]
> We have to document that the user should install chronyd before starting.

@roydahan @fgelcer , according to this comment, the following changes are required:
1. Change Scylla offline install doc that user has a prerequisite to install either chrony/systemd-timesyncd/ntp.
2. Change the SCT validation to check that if none of the above is installed then SCT installs chrony.

please comment.

Note: the above install doc seems to be a private and not a scylla official one.

**Comment by fgelcer** [comment]
> > We have to document that the user should install chronyd before starting.
> 
> @roydahan @fgelcer , according to this comment, the following changes are required:
> 
> 1. Change Scylla offline install doc that user has a prerequisite to install either chrony/systemd-timesyncd/ntp.
> 2. Change the SCT validation to check that if none of the above is installed then SCT installs chrony.
> 
> please comment.
> 
> Note: the above install doc seems to be a private and not a scylla official one.

@yarongilor , work with @annastuchlik and find the document for offline install, and check what are the dependencies... AFAICT, the implementation in SCT followed the recommendations at the time, although it maybe didn't make to the official documentation of ScyllaDB.
if that is the case, and we need to install in advance any package, we need to re-work the SCT code, as the verification it is doing is wrong.

**Comment by yarongilor** [comment]
> > We have to document that the user should install chronyd before starting.
> 
> @roydahan @fgelcer , according to this comment, the following changes are required:
> 
> 1. Change Scylla offline install doc that user has a prerequisite to install either chrony/systemd-timesyncd/ntp.
> 2. Change the SCT validation to check that if none of the above is installed then SCT installs chrony.
> 
> please comment.
> 
> Note: the above install doc seems to be a private and not a scylla official one.

@syuu1228 , can you please advise - 
1. is this requirement correct?
2. Where can we find an updated install guide doc with the specified requirements?

**Comment by avikivity** [comment]
We can't bring every single dependency. 

The installer should just skip chrony/ntp if not available and move on, with a warning.

**Comment by yarongilor** [comment]
OK so we can close this issue and have the following update to SCT:
1. Non-offline-install tests - left as-is with same validations.
2. Offline-install tests - the validation for tyme-sync services should be changed the following:
if no time-sync service exist on node - then the offline-install script should print out a warning for that during its install process. SCT can check if such a warning exist in installation output.

Cc @ShlomiBalalis , @roydahan

**Comment by yarongilor** [comment]
> > systemd-timesyncd is not installed on oel76 instance during artifacts test, yet Scylla did not install ntp or chrony services as a replacement
> 
> scylla_ntp_setup will install chrony automatically on RHEL variants. See the logic on the script: https://github.com/scylladb/scylla/blob/next/dist/common/scripts/scylla_ntp_setup#L55 It will automatically install chrony on RHEL variants, and install systemd-timesyncd on other distros.
> 
> Why SCT log is saying "did not install ntp or chrony"?

@syuu1228 , it works in non-offline but doesn't work in offline mode, as detailed above - https://github.com/scylladb/scylla/issues/10608#issuecomment-1133928417

please advise if it looks a script issue that should be fixed.

**Comment by ShlomiBalalis** [comment]
> > systemd-timesyncd is not installed on oel76 instance during artifacts test, yet Scylla did not install ntp or chrony services as a replacement
> 
> scylla_ntp_setup will install chrony automatically on RHEL variants. See the logic on the script: https://github.com/scylladb/scylla/blob/next/dist/common/scripts/scylla_ntp_setup#L55 It will automatically install chrony on RHEL variants, and install systemd-timesyncd on other distros.
> 
> Why SCT log is saying "did not install ntp or chrony"?

@syuu1228 We check if systemd-timesync is installed, and if not we expect scylla to install either ntp or chrony. In this scenario, none of the services existed after the scylla installation was complete.
Avi says that in offline installations, scylla does not take care of this dependency and it's the user's responsibility, so we should not run this test on offline installation scenarios

**Comment by rayakurl** [comment]
@syuu1228 - can you please reply to @ShlomiBalalis comment above? thanks

**Comment by syuu1228** [comment]
> > > systemd-timesyncd is not installed on oel76 instance during artifacts test, yet Scylla did not install ntp or chrony services as a replacement
> > 
> > 
> > scylla_ntp_setup will install chrony automatically on RHEL variants. See the logic on the script: https://github.com/scylladb/scylla/blob/next/dist/common/scripts/scylla_ntp_setup#L55 It will automatically install chrony on RHEL variants, and install systemd-timesyncd on other distros.
> > Why SCT log is saying "did not install ntp or chrony"?
> 
> @syuu1228 , it works in non-offline but doesn't work in offline mode, as detailed above - [#10608 (comment)](https://github.com/scylladb/scylla/issues/10608#issuecomment-1133928417)
> 
> please advise if it looks a script issue that should be fixed.

Ah, so what you are testing is make sure offline-installer configures NTP same as normal installation (.rpm/.deb)?
Unfortunately it will skip configuring NTP, and it's expected behavior.

Because, on offline environment it unlikely able to connect public NTP server on the internet, we cannot use default configuration on NTP clients (ntpd or chrony or systemd-timesyncd).
On such environment, the user probably have NTP server on Local network, but we won't able to auto-detect it, so user has responsibility to configure it manually, so we just skip running NTP configuration on offline-installer.

Probably we can add an option to specifying local NTP server, though.


(BTW, the test is disabling or uninstalling chronyd before running scylla_setup, right? Looks like chronyd.service is actually running by default on the AMI)

**Comment by roydahan** [comment]
@yarongilor till @ShlomiBalalis will reach it may take time.
Can you please send the fix yourself?

**Comment by yarongilor** [comment]
> @yarongilor till @ShlomiBalalis will reach it may take time. Can you please send the fix yourself?

@roydahan , i think there's already PR https://github.com/scylladb/scylla-cluster-tests/pull/4792 that we can merge.
I can open a task for the 'TODO' part left (probably with a lower priority).
